crates/doc/src/shape/limits.rs

```diff
@@ -1,63 +1,180 @@
-        return;
+fn squash_addl_properties(props: Option<Box<Shape>>) -> Option<Box<Shape>> {
+    match props {
+        Some(inner) if inner.type_.eq(&types::INVALID) => Some(Box::new(Shape::nothing())),
+        Some(_) => Some(Box::new(Shape::anything())),
+        None => None,
     }
+}
 
-    let limit = match loc {
-        Location::Root => MAX_ROOT_FIELDS,
-        _ => MAX_NESTED_FIELDS,
-    };
+// Squashing a shape inside an array tuple is special because the location
+// of shapes inside the tuple is _itself_ the key into that container.
+// This means that if we do anything to shift the keys of still-existing shapes,
+// they won't be valid any longer. With that in mind, there's also no good reason
+// to squash one object field over any other, so let's just treat
+// Token::Index and Token::Property as signals to squash _an_ index or property,
+// leaving it up to the implementation to determine which one.
+fn squash_location_inner(shape: &mut Shape, name: &Token) {
+    match name {
+        // Squashing of `additional*` fields is not possible here because we don't
+        // have access to the parent shape
+        Token::NextIndex => unreachable!(),
+        Token::Property(prop) if prop == "*" => unreachable!(),
 
-    if slf.object.properties.len() > limit {
-        // Take all of the properties' shapes and
-        // union them into additionalProperties
+        Token::Index(_) => {
+            // Remove the last location from the array tuple shape
+            let mut shape_to_squash = shape
+                .array
+                .tuple
+                .pop()
+                .expect("No array tuple property to squash");
 
-        let existing_additional_properties = slf
-            .object
-            .additional_properties
-            // `Shape::union` takes owned Shapes which is why we
-            // have to take ownership here.
-            .take()
-            .map(|boxed| *boxed)
-            .unwrap_or(Shape::nothing());
-
-        let merged_additional_properties = slf
-            .object
-            .properties
-            // As part of squashing all known property shapes together into
-            // additionalProperties, we need to also remove those explicit properties.
-            .drain(..)
-            .fold(existing_additional_properties, |accum, mut prop| {
-                // Recur here to avoid excessively large `additionalProperties` shapes
-                enforce_field_count_limits(&mut prop.shape, loc.push_prop(&prop.name));
-                Shape::union(accum, prop.shape)
-            });
-
-        slf.object.additional_properties = Some(Box::new(merged_additional_properties));
-    } else {
-        for prop in slf.object.properties.iter_mut() {
-            enforce_field_count_limits(&mut prop.shape, loc.push_prop(&prop.name))
+            shape_to_squash.array.additional_items =
+                squash_addl_properties(shape_to_squash.array.additional_items);
```

Comment:
can't you just set `additional_items` and `additional_properties` to `None` ?

nit: if you can't and must keep this, then `squash_addl_properties` => `squash_addl` ? sometimes item, sometimes property